packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,548 @@
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import React, {createContext, ForwardedRef, forwardRef, HTMLAttributes, JSX, ReactElement, ReactNode, RefObject, useContext, useEffect, useMemo, useRef} from 'react';
+import {useControlledState} from '@react-stately/utils';
+
+class TreeCollection<T> implements ICollection<Node<T>> {
+  private flattenedRows: Node<T>[];
+  private keyMap: Map<Key, NodeValue<T>> = new Map();
+
+  constructor(opts) {
+    let {collection, expandedKeys} = opts;
+    let {flattenedRows, keyMap} = flattenTree<T>(collection, {expandedKeys});
+    this.flattenedRows = flattenedRows;
+    // Use generated keyMap because it contains the modified collection nodes (aka it adjusts the indexes so that they ignore the existence of the Content items)
+    this.keyMap = keyMap;
+  }
+
+  // TODO: should this collection's getters reflect the flattened structure or the original structure
+  // If we respresent the flattened structure, it is easier for the keyboard nav but harder to find all the nodes
+  *[Symbol.iterator]() {
+    yield* this.flattenedRows;
+  }
+
+  get size() {
+    return this.flattenedRows.length;
+  }
+
+  getKeys() {
+    return this.keyMap.keys();
+  }
+
+  getItem(key: Key): Node<T> | null {
+    return this.keyMap.get(key) || null;
+  }
+
+  at(idx: number) {
+    return this.flattenedRows[idx];
+  }
+
+  getFirstKey() {
+    return this.flattenedRows[0]?.key;
+  }
+
+  getLastKey() {
+    return this.flattenedRows[this.size - 1]?.key;
+  }
+
+  getKeyAfter(key: Key) {
+    let index = this.flattenedRows.findIndex(row => row.key === key);
+    return this.flattenedRows[index + 1]?.key;
+  }
+
+  getKeyBefore(key: Key) {
+    let index = this.flattenedRows.findIndex(row => row.key === key);
+    return this.flattenedRows[index - 1]?.key;
+  }
+
+  // Note that this will return Content nodes in addition to nested TreeItems
+  getChildren(key: Key): Iterable<Node<T>> {
+    let keyMap = this.keyMap;
+    return {
+      *[Symbol.iterator]() {
+        let parent = keyMap.get(key);
+        let node = parent?.firstChildKey != null ? keyMap.get(parent.firstChildKey) : null;
+        while (node) {
+          yield node as Node<T>;
+          node = node.nextKey != null ? keyMap.get(node.nextKey) : undefined;
+        }
+      }
+    };
+  }
+
+  getTextValue(key: Key): string {
+    let item = this.getItem(key);
+    return item ? item.textValue : '';
+  }
+}
+
+export interface TreeRenderProps {
+  /**
+   * Whether the tree has no items and should display its empty state.
+   * @selector [data-empty]
+   */
+  isEmpty: boolean,
+  /**
+   * Whether the tree is currently focused.
+   * @selector [data-focused]
+   */
+  isFocused: boolean,
+  /**
+   * Whether the tree is currently keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean,
+  /**
+   * State of the grid list.
```

Comment:
```suggestion
   * State of the tree.
```

Reply:
urgh, thought I got fixed all of those thanks